lib/include/jxl/decode.h

```diff
@@ -746,6 +746,56 @@ JXL_EXPORT JxlDecoderStatus JxlDecoderGetColorAsICCProfile(
 JXL_EXPORT JxlDecoderStatus JxlDecoderSetPreferredColorProfile(
     JxlDecoder* dec, const JxlColorEncoding* color_encoding);
 
+/**
+ * Sets the color management system (CMS) that will be used for color conversion
+ * (if applicable) during decoding. May only be set before starting decoding. If
+ * left unset, the decoder will not do arbitrary color conversions:
+ * JxlDecoderSetPreferredColorProfile can be used, but there is no guarantee
+ * that the decoded image will be returned in the preferred color encoding. If
+ * however a CMS is set using this function, then the function
+ * JxlDecoderSetColorEncoding or JxlDecoderSetICCProfile can be used and the
+ * decoded image will then be guaranteed to be in the requested color encoding.
+ *
+ * @param dec decoder object.
+ * @param cms structure representing a CMS implementation. See JxlCmsInterface
+ * for more details.
+ */
+JXL_EXPORT void JxlDecoderSetCms(JxlDecoder* dec, JxlCmsInterface cms);
+
+/**
+ * Sets the desired output color encoding of the decoded image. This function
+ * can only be used if JxlDecoderSetCms was previously set. This is an
+ * alternative to JxlDecoderSetICCProfile and only one of these two must be
+ * used. This one sets the color encoding as a @ref JxlColorEncoding, while the
+ * other sets it as ICC binary data.
+ *
+ * @param dec decoder object.
+ * @param color color encoding. Object owned by the caller and its contents are
+ * copied internally.
+ * @return JXL_DEC_SUCCESS if the operation was successful, JXL_DEC_ERROR
+ * otherwise.
+ */
+JXL_EXPORT JxlDecoderStatus
+JxlDecoderSetColorEncoding(JxlDecoder* dec, const JxlColorEncoding* color);
```

Comment:
How does this interact with SetPreferredColorEncoding?

Reply:
I'd say SetPreferred only does a conversion if it can do that without calling cms (even when a cms has been set), so lossless arbitrary colorspace remains untouched (including in particular CMYK, where conversion to RGB is particularly lossy).

Reply:
What happens if you call both?

In general, I think we should better document when to use one and when to use the other.

@sboukortt should we also mention the desired intensity target, perhaps?

Reply:
I added a paragraph of documentation to JxlDecoderSetCms to try to clarify things.

Desired intensity target can currently not be set, you only get the intensity target of the image from the basicinfo and you'll get output with the same intensity target, but probably it would be useful to also be able to set a desired intensity target for the output so applications don't have to do re-scaling themselves (in particular if they want to decode HDR images to SDR uint).

Reply:
Desired intensity target sounds like a bit of a pain (it means we would probably have to do some tone mapping, and likely different for HLG content), but perhaps it could be good to have indeed.

Reply:
as discussed below, settling for not having `JxlDecoderSetColorEncoding`, but letting `JxlDecoderSetPreferredIntensityTarget` take that role whether or not a CMS